File: hw1abc/abc/container.cpp
#include "container.h"

//------------------------------------------------------------------------------
// Инициализация контейнера.
void Init(container &c) {
    c.len = 0;
}

//------------------------------------------------------------------------------
// Очистка контейнера от элементов (освобождение памяти).
void Clear(container &c) {
    for (int i = 0; i < c.len; ++i) {
        delete c.cont[i];
    }
    c.len = 0;
}

//------------------------------------------------------------------------------
// Ввод содержимого контейнера из указанного потока.
void In(container &c, FILE *file) {
    while(!feof(file)) {
        if((c.cont[c.len] = In(file)) !=  0) {
            c.len++;
        }
    }
}

//------------------------------------------------------------------------------
// Случайный ввод содержимого контейнера.
void InRandom(container &c, int size) {
    while(c.len < size) {
        if((c.cont[c.len] = InRandom()) != 0) {
            c.len++;
        }
    }
}

//------------------------------------------------------------------------------
// Вывод содержимого контейнера в указанный поток.
void Out(container &c, FILE *file) {
    fprintf(file, "Container contains %d elements.\n", c.len);
    for (int i = 0; i < c.len; i++) {
        fprintf(file, "%d: ", i);
        Out(*c.cont[i], file);
        fprintf(file, "%s", "\n");
    }
}

//------------------------------------------------------------------------------
// Вычисление среднего арифметического средних арифметических всех матриц в контейнере.
double ArithmeticMeanAll(container &c) {
    double sum = 0.0;
    for(int i = 0; i < c.len; i++) {
        sum += ArithmeticMean(*c.cont[i]);
    }
    return sum;
}

//------------------------------------------------------------------------------
// Сортировка контейнера по убыванию методом Шелла.
void ShellSorting(container &c){
    int step = c.len / 2;
    matrix* temp;
    while (step > 0) {
        for (int i = 0; i < (c.len - step); ++i) {
            int j = i;
            while (j >= 0 && ArithmeticMean(*c.cont[j]) < ArithmeticMean(*c.cont[j + step])) {
                temp = c.cont[j];
                c.cont[j] = c.cont[j + step];
                c.cont[j + step] = temp;
                j -= step;
            }
        }
        step /= 2;
    }
}

File: hw3abc/abc3/container.py
import random
from matrix import Matrix
from twoDimensionalArray import TwoDimensionalArray
from diagonalMatrix import DiagonalMatrix
from lowerTriangularMatrix import LowerTriangularMatrix


# ------------------------------------------------------------------------------
# container.py - содержит тип данных,
# представляющий простейший контейнер
# ------------------------------------------------------------------------------

# ------------------------------------------------------------------------------
# Простейший контейнер на основе одномерного массива.
class Container:
    # Текущая длина.
    __size__: int
    __storage__: list

    # Конструктор контейнера.
    def __init__(self, size: int):
        self.__size__ = size
        self.__storage__ = []

    # Ввод содержимого контейнера из указанного файла.
    def read_file(self, lines: list):
        counter = 0
        for i in range(0, self.__size__):
            matrix_type = int(lines[counter])
            counter += 1
            num = int(lines[counter])
            counter += 1

            if matrix_type == 1:
                m = TwoDimensionalArray(num)
                m.read_file(lines[counter])
                self.__storage__.append(m)
                counter += 1
            elif matrix_type == 2:
                dm = DiagonalMatrix(num)
                dm.read_file(lines[counter])
                self.__storage__.append(dm)
                counter += 1
            else:
                ltm = LowerTriangularMatrix(num)
                ltm.read_file(lines[counter])
                self.__storage__.append(ltm)
                counter += 1

    # Случайный ввод содержимого контейнера.
    def random_read(self):
        for i in range(0, self.__size__):
            matrix_type = random.randint(0, 3)

            if matrix_type == 0:
                m = TwoDimensionalArray(0)
                m.random_read()
                self.__storage__.append(m)
            elif matrix_type == 1:
                dm = DiagonalMatrix(0)
                dm.random_read()
                self.__storage__.append(dm)
            else:
                ltm = LowerTriangularMatrix(0)
                ltm.random_read()
                self.__storage__.append(ltm)

    # Вывод содержимого контейнера в указанный файл.
    def write(self) -> str:
        out_str = "Container contains " + str(self.__size__) + " elements.\n"
        for item in self.__storage__:
            out_str += str(item.write())

        return out_str

    # Сортировка Шелла.
    def shell_sorting(self):
        length = self.__size__
        step = length // 2
        while step > 0:
            for i in range(step, length, 1):
                temp = self.__storage__[i]
                j = i
                while (j >= step) & (self.__storage__[j - step].arithmeticMean() < temp.arithmeticMean()):
                    self.__storage__[j] = self.__storage__[j - step]
                    j -= step
                self.__storage__[j] = temp
            step = step // 2
        pass
